### PR TITLE
Multistore ShopUuid manager and Multistore fixer

### PR DESCRIPTION
**Pull request by Matt75**
Please note Configuration::get return value for all shops or for group if no value for shop is found

### CODE REVIEW

File: classes/MultiStoreFixer.php [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+        foreach (self::UNIQUE_CONFIGURATION_KEYS as $key) {
+            // Get current value
+            $value = \Configuration::get($key);
+
+            // Remove current global value
+            \Configuration::deleteByName($key);
+
+            // Save current value associated to default shop
+            $result &= \Configuration::updateValue(
```
Comment (Quetzacoalt91) [inline classes/MultiStoreFixer.php]: The operator `&=` is actually a bitwise operator, which returns an integer.

When you deal with booleans, you should use the following syntax:

```suggestion
            $result = $result && \Configuration::updateValue(
```

This allows you to remove the `(bool)` at the end of the method.

File: classes/MultiStoreFixer.php [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+                $shopHasUniqueKey = \Configuration::hasKey($key, null, null, (int) $shopId);
+                $shopValue = \Configuration::get($key, null, null, (int) $shopId);
+                $shopHasValue = false === empty($shopValue);
+                $isDefaultShop = (int) $shopId === (int) \Configuration::get('PS_SHOP_DEFAULT');
+
+                if ($shopHasUniqueKey && $shopHasValue) {
+                    // This shop has a unique value, we keep it
+                    $currentValuesByShop[(int) $shopId][$key] = $shopValue;
+                } elseif (false === $globalValueUsed && $shopHasValue && in_array($shopId, $shopListUsedByOrder)) {
+                    // This shop has order placed with checkout, we take global value found (can be value of group or all shop)
```
Comment (Quetzacoalt91) [inline classes/MultiStoreFixer.php]: Are you sure about this comment? I don't get the relation between the configuration data & an order.

File: classes/MultiStoreFixer.php [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+            }
+        }
+
+        foreach (self::UNIQUE_CONFIGURATION_KEYS as $key) {
+            // Remove values
+            \Configuration::deleteByName($key);
+
+            // Save values for each shop
+            foreach (\Shop::getShops(false, null, true) as $shopId) {
+                $result &= \Configuration::updateValue(
```
Comment (Quetzacoalt91) [inline classes/MultiStoreFixer.php]: Same comment about the bitwise operator and booleans

File: classes/ShopUuidManager.php [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+     *
+     * @return bool
+     */
+    public function generateForShop($idShop)
+    {
+        $result = true;
+
+        if (false === \Configuration::hasKey('PS_CHECKOUT_SHOP_UUID_V4', null, null, (int) $idShop) || !$this->getForShop($idShop)) {
+            $uuid4 = \Ramsey\Uuid\Uuid::uuid4();
+            $result &= \Configuration::updateValue(
```
Comment (Quetzacoalt91) [inline classes/ShopUuidManager.php]:
```suggestion
            $result = $result &&  \Configuration::updateValue(
```

File: classes/ShopUuidManager.php [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                'PS_CHECKOUT_SHOP_UUID_V4',
+                $uuid4->toString(),
+                false,
+                null,
+                (int) $idShop
+            );
+        }
+
+        return (bool) $result;
```
Comment (Quetzacoalt91) [inline classes/ShopUuidManager.php]:
```suggestion
        return $result;
```

File: classes/ShopUuidManager.php [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+     * Used in module installation
+     *
+     * @return bool
+     */
+    public function generateForAllShops()
+    {
+        $result = true;
+
+        foreach (\Shop::getShops(false, null, true) as $shopId) {
+            $result &= $this->generateForShop($shopId);
```
Comment (Quetzacoalt91) [inline classes/ShopUuidManager.php]:
```suggestion
            $result = $result && $this->generateForShop($shopId);
```

File: ps_checkout.php [hunk 7]
```diff
@@ -167,25 +168,23 @@ public function install()
...
-        $shopsList = \Shop::getShops(false, null, true);
+        $result = true;
 
-        foreach ($shopsList as $shopId) {
+        foreach (\Shop::getShops(false, null, true) as $shopId) {
             foreach ($this->configurationList as $name => $value) {
-                if ($name === 'PS_CHECKOUT_SHOP_UUID_V4') {
-                    $value = $uuid4->toString();
+                if (false === Configuration::hasKey($name, null, null, (int) $shopId)) {
+                    $result &= Configuration::updateValue(
```
Comment (Quetzacoalt91) [inline ps_checkout.php]:
```suggestion
                    $result = $result && Configuration::updateValue(
```

File: classes/MultiStoreFixer.php (hunk 2 above)
Comment (Matt75) [reply classes/MultiStoreFixer.php]: Yes, thanks to `in_array($shopId, $shopListUsedByOrder)` I check if this shop has been used to place an order.

File: ps_checkout.php [hunk 8]
```diff
@@ -173,7 +173,7 @@ public function installConfiguration()
         foreach (\Shop::getShops(false, null, true) as $shopId) {
             foreach ($this->configurationList as $name => $value) {
                 if (false === Configuration::hasKey($name, null, null, (int) $shopId)) {
-                    $result &= Configuration::updateValue(
+                    $result && Configuration::updateValue(
```
Comment (Quetzacoalt91) [inline ps_checkout.php]: Be careful, you forgot to assign the result in `$result`.

**Pull request merged**